JujuHooks/JujuHooks.psm1

```diff
@@ -0,0 +1,1225 @@
+     to happen, and will wait for any currently running hook to finish before acquiring the execution lock
+     and executing a reboot. It also waits for any running container to shutdown.
+    .PARAMETER Now
+     This parameter makes the Invoke-JujuReboot commandlet to block until reboot is executed. It guarantees
+     that if executed with the -Now parameter, your script will not continue after that point. You should take
+     great care when using this option, as you might enter into a reboot loop.
+     Omitting this option will schedule a reboot at the end of the currently running hook.
+    .EXAMPLE
+     # Check if feature is installed
+     $isInstalled = Check-IfFeatureIsInstalled
+     if(!$isInstalled){
+        # Feature is not installed.
+        # Its important to suppress any automatic reboot from any commandlet. Rebooting
+        # a server while the hook is running will error out the hook.
+        $result = Install-FeatureThatRequiresReboot -Reboot:$False
+
+        # Some commandlets return a result that contains whether or not a reboot is required
+        # before changes take effect
+        if($result.RebootRequired){
+            # Only reboot if we absolutely have to.
+            Invoke-JujuReboot -Now
+        }
+     }  
+    #>
+    Param(
+        [switch]$Now
+    )
+    $cmd = @("juju-reboot.exe")
+
+    if ($Now) {
+        $cmd += "--now"
+    }
+    Invoke-JujuCommand -Command $cmd | Out-Null
+}
+
+# TODO(gabriel-samfira): Move this to separate module
+function Get-MainNetadapter {
+    <#
+    .SYNOPSIS
+    Returns the interface alias of the primary network adapter. The primary network adapter in this
+    case is the NIC that has the IP address that juju is aware of, configured. So if the IP address
+    returned by Get-JujuUnitPrivateIP is configured on a NIC, that NIC is the primary one.
+    #>
+    PROCESS {
+        $unit_ip = Get-JujuUnitPrivateIP
+        if (!$unit_ip) {
+            Throw "Failed to get unit IP"
+        }
+
+        $iface = Get-NetIPAddress | Where-Object { $_.IPAddress -match $unit_ip -and $_.AddressFamily -eq "IPv4" }
+        if ($iface) {
+            $ifaceAlias = $iface.InterfaceAlias
+            if ($ifaceAlias) {
+                return $ifaceAlias
+            } else {
+                Throw "Interface alias is null."
+            }
+        } else {
+            Throw "Failed to find primary interface."
+        }
+    }
+}
+
+function Get-PrimaryAdapterDNSServers {
+    <#
+    .SYNOPSIS
+    Returns the DNS servers configured for the primary network adapter. See Get-MainNetadapter
+    for a definition of "primary network adapter"
+    #>
+    $netAdapter = Get-MainNetadapter
+    $dnsServers = (Get-DnsClientServerAddress -InterfaceAlias $netAdapter `
+                  -AddressFamily IPv4).ServerAddresses
+    return $dnsServers
+}
+
+function Confirm-JujuPortRangeOpen {
```

Comment:
I think that the issue with overlapping IPs reported [here](https://github.com/cloudbase/juju-powershell-modules/issues/2) still applies here.